File: AgoraChat/Code/Controllers/Groups/ACDGroupMemberBlockListViewController.m
//
//  ACDGroupMemberBlockViewController.m
//  ChatDemo-UI3.0
//
//  Created by liang on 2021/10/29.
//  Copyright © 2021 easemob. All rights reserved.
//

#import "ACDGroupMemberBlockListViewController.h"
#import "ACDContactCell.h"
#import "UIViewController+HUD.h"
#import "AgoraNotificationNames.h"
#import "ACDContainerSearchTableViewController+GroupMemberList.h"
#import "AgoraUserModel.h"

@interface ACDGroupMemberBlockListViewController ()

@property (nonatomic, strong) AgoraChatGroup *group;

@end

@implementation ACDGroupMemberBlockListViewController

- (instancetype)initWithGroup:(AgoraChatGroup *)aGroup
{
    self = [super init];
    if (self) {
        self.group = aGroup;
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(updateGroupMemberWithNotification:) name:KACD_REFRESH_GROUP_MEMBER object:nil];
    }
    
    return self;
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}


- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self useRefresh];
    
    if (self.group.permissionType == AgoraChatGroupPermissionTypeOwner || self.group.permissionType == AgoraChatGroupPermissionTypeAdmin) {
        [self tableViewDidTriggerHeaderRefresh];
    }
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

#pragma mark refresh and load more
- (void)didStartRefresh {
    [self tableViewDidTriggerHeaderRefresh];
}

- (void)didStartLoadMore {
    [self tableViewDidTriggerFooterRefresh];
}



#pragma mark NSNotification
- (void)updateGroupMemberWithNotification:(NSNotification *)aNotification {
    NSDictionary *dic = (NSDictionary *)aNotification.object;
    NSString* groupId = dic[kACDGroupId];
    ACDGroupMemberListType type = [dic[kACDGroupMemberListType] integerValue];
    
    if (![self.group.groupId isEqualToString:groupId] || type != ACDGroupMemberListTypeBlock) {
        return;
    }
    
    [self tableViewDidTriggerHeaderRefresh];
}


#pragma mark - Table view data source
- (NSInteger)numberOfSectionsInTableView:(UITableView *)table {
    if (self.isSearchState) {
        return 1;
    }
    return  self.sectionTitles.count;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    if (self.isSearchState) {
        return self.searchResults.count;
    }
    return ((NSArray *)self.dataArray[section]).count;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return [ACDContactCell height];
}
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    ACDContactCell *cell = (ACDContactCell *)[tableView dequeueReusableCellWithIdentifier:[ACDContactCell reuseIdentifier]];
    if (cell == nil) {
        cell = [[ACDContactCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:[ACDContactCell reuseIdentifier]];
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
    }
    
    AgoraUserModel *model = nil;
    if (self.isSearchState) {
        model = self.searchResults[indexPath.row];
    }else {
        model = self.dataArray[indexPath.section][indexPath.row];
    }

    cell.model = model;
    
    ACD_WS
    cell.tapCellBlock = ^{
        [weakSelf actionSheetWithUserId:model.hyphenateId memberListType:ACDGroupMemberListTypeBlock group:weakSelf.group];
    };
    
    return cell;
}


- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
}

#pragma mark - data

- (void)tableViewDidTriggerHeaderRefresh
{
    self.page = 1;
    [self fetchBlocksWithPage:self.page isHeader:YES];
}

- (void)tableViewDidTriggerFooterRefresh
{
    self.page += 1;
    [self fetchBlocksWithPage:self.page isHeader:NO];
}

- (void)fetchBlocksWithPage:(NSInteger)aPage
                 isHeader:(BOOL)aIsHeader
{
    NSInteger pageSize = 50;
    ACD_WS

    [[AgoraChatClient sharedClient].groupManager getGroupBlacklistFromServerWithId:self.group.groupId pageNumber:self.page pageSize:pageSize completion:^(NSArray *aMembers, AgoraChatError *aError) {
        
        [self endRefresh];
        if (!aError) {
            [self updateUIWithResultList:aMembers IsHeader:aIsHeader];
        } else {
            NSString *errorStr = [NSString stringWithFormat:NSLocalizedString(@"group.ban.fetchFail", @"Fail to get blacklist: %@"), aError.errorDescription];
            [weakSelf showHint:errorStr];
        }
        
        if ([aMembers count] < pageSize) {
            [self endLoadMore];
        } else {
            [self useLoadMore];
        }
    }];
}

- (void)updateUIWithResultList:(NSArray *)sourceList IsHeader:(BOOL)isHeader {
    
    if (isHeader) {
        [self.members removeAllObjects];
    }
    [self.members addObjectsFromArray:sourceList];
    
    [self sortContacts:self.members];

    dispatch_async(dispatch_get_main_queue(), ^(){
        [self.table reloadData];
    });
}

@end

File: AgoraChat-Swift/AgoraChat-Swift/code/Controllers/Groups/ACDGroupMemberBlockListViewController.swift
//
//  ACDGroupMemberBlockListViewController.swift
//  AgoraChat-Swift
//
//  Created by 冯钊 on 2022/10/21.
//

import UIKit

class ACDGroupMemberBlockListViewController: ACDContainerSearchTableViewController<[AgoraUserModel]> {

    private var group: AgoraChatGroup
    
    init(group: AgoraChatGroup) {
        self.group = group
        super.init(nibName: nil, bundle: nil)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        NotificationCenter.default.addObserver(self, selector: #selector(updateGroupMemberWithNotification(_:)), name: GroupMemberChangedNotification, object: nil)
        
        self.table.rowHeight = 54
        self.table.register(UINib(nibName: "ACDContactCell", bundle: nil), forCellReuseIdentifier: "cell")
        
        self.useRefresh()
        if self.group.permissionType == .owner || self.group.permissionType == .admin {
            self.updateUI()
        }
    }
    
    func updateUI() {
        if self.group.permissionType == .owner || self.group.permissionType == .admin {
            self.fetchBlocks(isRefresh: true)
        }
    }
    
    private func fetchBlocks(isRefresh: Bool) {
        if isRefresh {
            self.page = 1
        } else {
            self.page += 1
        }
        AgoraChatClient.shared().groupManager?.getGroupBlacklistFromServer(withId: self.group.groupId, pageNumber: Int(self.page), pageSize: 20, completion: { list, error in
            self.endRefresh()
            if let error = error {
                self.showHint(error.errorDescription)
            } else if let list = list {
                if (isRefresh) {
                    self.members.removeAll()
                }
                self.members.append(contentsOf: list)
                self.sortAndReloadContacts(self.members)
            }
            if list?.count ?? 0 < 20 {
                self.endLoadMore()
            } else {
                self.useLoadMore()
            }
        })
    }
    
    @objc private func updateGroupMemberWithNotification(_ notification: Notification) {
        guard let dict = notification.object as? [String: Any] else {
            return
        }
        if self.group.groupId == dict["kACDGroupId"] as? String && dict["kACDGroupMemberListType"] as? ACDContainerSearchTableViewController<[AgoraUserModel]>.GroupMemberShowType == .block {
            self.updateUI()
        }
    }
    
    override func didStartRefresh() {
        self.fetchBlocks(isRefresh: true)
    }
    
    override func didStartLoadMore() {
        self.fetchBlocks(isRefresh: false)
    }
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return self.isSearchState ? 1 : self.sectionTitles.count
    }
    
    func sectionIndexTitles(for tableView: UITableView) -> [String]? {
        return self.sectionTitles
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.isSearchState ? self.searchResults.count : self.dataArray[section].count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath)
        if let cell = cell as? ACDContactCell {
            var model: AgoraUserModel?
            if self.isSearchState {
                model = self.searchResults[indexPath.row] as? AgoraUserModel
            } else {
                model = self.dataArray[indexPath.section][indexPath.row]
            }
            if let model = model {
                cell.model = model
                cell.tapCellHandle = { [unowned self] in
                    self.showActionSheet(userId: model.hyphenateId, showType: .block, group: self.group)
                }
            }
        }
        return cell
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }
}
